src/Functions/IFunctionDateOrDateTime.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDate32.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <Functions/IFunction.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <Functions/extractTimeZoneFromFunctionArguments.h>
+#include <Functions/DateTimeTransforms.h>
+#include <Functions/TransformDateTime64.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/Context.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+template <typename Transform>
+class IFunctionDateOrDateTime : public IFunction
```

Comment:
Too generic name. But it is hard to find a better name. `IFunctionWithDateDate32DateTimeOrDateTime64Argument` would be correct but too long.

Just as an idea ... what about
- removing this base class altogether and
- keeping `FunctionDateOrDateTimeToSomething` as it was before and
- building `FunctionDateOrDateTimeToDateOrDate32` from `FunctionDateOrDateTimeToSomething` as a template (i.e. using some copypasta) but with `enable_date32_results` in mind?

If I understand correctly, the final goal is to convert *all* functions which currently use `FunctionDateOrDateTimeToSomething` to `FunctionDateOrDateOrDate32`. When this future goal is reached, `FunctionDateOrDateTimeToSomething` can go away anyways and we would end up with a weird class hierarchy (Function --> FunctionDateOrDateOrDate32 --> IFunctionDateOrDateTime --> IFunction). So the alternative is to accept some copypaste for a period of time but the end result is prettier.

Reply:
Robert, I think FunctionDateOrDateTimeToSomething should be kept forever. But it's not a good name. "Something" is confusing. Ideally it should be renamed into FunctionDateOrDateTimeTo**OneType**.
Because the "nature" of this class is to accept Date, Date32, DateTime, DateTime64 as **argument**, but always return just **one result** type. 
Consider:

src/Functions/toISOWeek.cpp
`using FunctionToISOWeek = FunctionDateOrDateTimeToSomething<DataTypeUInt8, ToISOWeekImpl>;`

toSOWeek function returns just one argument UInt8. 
Thus FunctionDateOrDateTimeToSomething must exist. It's not a "transitional"/temp entity.

Consider my class diagram:
![class diagram](https://user-images.githubusercontent.com/18441691/190195997-be9efe6b-bc51-4b1b-8dc1-5f6c1166b7b1.png)


In this case base class IFunctionDateOrDateTime makes a job of **arguments** verification
While the job of FunctionDateOrDateTimeToSomething or FunctionDateOrDateTimeToDateOrDate32 is to 
1. define **results** types 
2. call appropriate execute methods. 
So it's a power of OOP.

Another note: there are several functions returning DateTime. And there is the same problem, these functions must return DateTime or DateTime64. For this purpose in another PR I would implement new class FunctionDateOrDateTimeToDateTimeOrDateTime64. 
The advantage is that we don't put "all eggs into one basket" ("class"). And there is more clear code separation. C++ files are not big and easy to read. I think it's more in ClickHouse-way: for the same reason we have many *MergeTree engines.

Reply:
Okay, if the functionality of `FunctionDateOrDateTimeToSomething` is still needed in future, then it must stay of course. But instead of naming it `FunctionDateOrDateTimeToOneType` let's name it just `FunctionDateOrDateTimeToType`. I think it's pretty clear from the code that "Type" is given by template parameter `ToType`.

About the new class hierarchy ... one could make the argument that even with OOP class hierarchies should be kept flat / not too deep. So an alternative design would be to remove "IFunctionDateOrDateTime.h" and to copy it's one-line functions into its three current subclasses and the longer functions as static methods into an external helper class which is referenced by the subclasses. Think about it if that design is simpler. But maybe it's a matter of personal preference. I am fine with either approach.

Reply:
To me `FunctionDateOrDateTimeToType` is arguable better than `FunctionDateOrDateTimeToSomething` indeed. But I am not sure that difference is significat enough to bother renaming.

Reply:
An alternative is to open a separate PR (maybe once this PR is in) where all the "noise" (renamings, refactorings) is done. That other PR will have a good chance to be merged quickly.

Reply:
To me probably in current PR it's better to keep FunctionDateOrDateTimeToSomething. And maybe rename it in future PR. Because, yes, there are many renamings already in current PR.